src/software/ai/passing/cost_function.cpp

```diff
@@ -8,70 +8,66 @@
 
-double ratePass(const World& world, const Pass& pass,
-                const std::optional<Rectangle>& target_region,
-                std::optional<unsigned int> passer_robot_id, PassType pass_type,
+double ratePass(const World& world, const Pass& pass, const Rectangle& zone,
                 std::shared_ptr<const PassingConfig> passing_config)
 {
     double static_pass_quality =
         getStaticPositionQuality(world.field(), pass.receiverPoint(), passing_config);
 
     double friendly_pass_rating = ratePassFriendlyCapability(
-        world.friendlyTeam(), pass, passer_robot_id, passing_config);
+        world.ball(), world.friendlyTeam(), pass, passing_config);
 
-    double enemy_pass_rating = ratePassEnemyRisk(world.enemyTeam(), pass, passing_config);
+    double enemy_pass_rating =
+        ratePassEnemyRisk(world.ball(), world.enemyTeam(), pass, passing_config);
 
-    double shoot_pass_rating =
-        ratePassShootScore(world.field(), world.enemyTeam(), pass, passing_config);
+    double shoot_pass_rating = ratePassShootScore(
+        world.ball(), world.field(), world.enemyTeam(), pass, passing_config);
 
-    // Rate all passes outside our target region as 0 if we have one
-    double in_region_quality = 1;
-    if (target_region)
-    {
-        in_region_quality = rectangleSigmoid(*target_region, pass.receiverPoint(), 0.1);
-    }
-
-    // Place strict limits on pass start time
-    double min_pass_time_offset =
-        passing_config->getMinTimeOffsetForPassSeconds()->value();
-    double max_pass_time_offset =
-        passing_config->getMaxTimeOffsetForPassSeconds()->value();
-    double pass_time_offset_quality =
-        sigmoid(pass.startTime().toSeconds(),
-                min_pass_time_offset + world.getMostRecentTimestamp().toSeconds(), 0.5) *
-        (1 - sigmoid(pass.startTime().toSeconds(),
-                     max_pass_time_offset + world.ball().timestamp().toSeconds(), 0.5));
+    double in_region_quality = rectangleSigmoid(zone, pass.receiverPoint(), 0.2);
 
     // Place strict limits on the ball speed
     double min_pass_speed     = passing_config->getMinPassSpeedMPerS()->value();
     double max_pass_speed     = passing_config->getMaxPassSpeedMPerS()->value();
     double pass_speed_quality = sigmoid(pass.speed(), min_pass_speed, 0.2) *
                                 (1 - sigmoid(pass.speed(), max_pass_speed, 0.2));
 
-    // We ignore shoot pass quality if we're rating for a "RECEIVE_AND_DRIBBLE" pass
-    double pass_quality = 0;
-    switch (pass_type)
+    return static_pass_quality * friendly_pass_rating * enemy_pass_rating *
+           shoot_pass_rating * pass_speed_quality * in_region_quality;
+}
+
+double rateZone(const Field& field, const Rectangle& zone, const Point& ball_position,
+                std::shared_ptr<const PassingConfig> passing_config)
+{
+    // If the zone contains the ball, its not a good place to receive a pass
+    if (contains(zone, ball_position))
```

Comment:
This generally makes sense, but technically doesn't scale well if you only divided the field into 4 zones for example. Ideally discouraging very short passes should be done implicitly by the other scoring functions.

Reply:
yeah that's true, this was more of a sanity check which doesn't really work for all pitch divisions

i've removed it, the other cost functions should already account for this